verilog/analysis/checkers/defparam_rule.cc

```diff
@@ -0,0 +1,68 @@
+// Copyright 2017-2019 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/defparam_rule.h"
+
+#include <string>
+#include <vector>
+
+#include "absl/strings/str_cat.h"
+#include "common/analysis/citation.h"
+#include "common/analysis/lint_rule_status.h"
+#include "common/analysis/matcher/bound_symbol_manager.h"
+#include "common/analysis/matcher/matcher.h"
+#include "common/text/symbol.h"
+#include "common/text/syntax_tree_context.h"
+#include "verilog/analysis/descriptions.h"
+#include "verilog/analysis/lint_rule_registry.h"
+
+namespace verilog {
+namespace analysis {
+
+using verible::GetStyleGuideCitation;
+
+// Register the lint rule
+VERILOG_REGISTER_LINT_RULE(DefparamRule);
+
+absl::string_view DefparamRule::Name() { return "forbid-defparam"; }
+const char DefparamRule::kTopic[] = "defparam";
+const char DefparamRule::kMessage[] =
+    "Do not use defparam.";
+
+std::string DefparamRule::GetDescription(
+    DescriptionType description_type) {
+  return absl::StrCat(
+      "Do not use defparam. See:",
+      GetStyleGuideCitation(kTopic), ".");
+}
+
+void DefparamRule::HandleSymbol(
+    const verible::Symbol& symbol, const verible::SyntaxTreeContext& context) {
+  verible::matcher::BoundSymbolManager manager;
+  if (matcher_.Matches(symbol, &manager)) {
+    if (!context.empty() && !context.top().children().empty()) {
+      auto defparam = context.top().children()[0].get();
```

Comment:
(See comment in .h file first on suggested node matcher.)
Take a look at GetSubtreeAsLeaf:
https://github.com/google/verible/blob/f74e900dbdd40c48fcaf1ff68db0c7f5bd342658/common/text/tree_utils.h#L102
With the update to the matcher, symbol will match at the node tagged `kParameterOverride`, whose child[0] will point to the `defparam` keyword leaf.  Then you don't need to manipulate the `context`.

```
const auto& defparam_token = GetSubtreeAsLeaf(symbol, NodeEnum::kParameterOverride, 0).get();
CHECK_EQ(defparam_token.token_enum, TK_defparam);
violations_.push_back(verible::LintViolation(defparam_token, kMessage, context));
```

Reply:
Right, that looks much cleaner this way, thanks for the detailed explanation!